File: SexyBiscuit.Engine/Rendering/Camera2D.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using SexyBiscuit.Engine.Core;
using SexyBiscuit.Engine.UI;   // RectangleF

namespace SexyBiscuit.Engine.Rendering;

/// <summary>
/// 2D camera component. Attach to an Actor to define the viewpoint for a scene.
/// Supports zoom, world-bounds clamping, smooth follow with deadzone, and trauma-based shake.
/// </summary>
public class Camera2D : Component
{
    // -------------------------------------------------------------------------
    // Which camera the 2D pass draws through
    // -------------------------------------------------------------------------

    private static readonly List<Camera2D> _all = new();

    /// <summary>
    /// The first active camera whose actor is tagged "MainCamera", else the first active one
    /// at all.
    /// </summary>
    /// <remarks>
    /// Camera3D has had this since it existed; Camera2D had no equivalent, which is part of
    /// why the 2D pass rendered with no camera at all and every camera-follow script looked
    /// broken in a native build while working in the browser. The fallback matches the
    /// browser's <c>Camera2D.main</c>: a scene with one untagged camera still draws through
    /// it rather than through nothing.
    /// </remarks>
    public static Camera2D? Main
    {
        get
        {
            foreach (var camera in _all)
                if (camera.Actor is { IsActive: true, Tag: "MainCamera" }) return camera;

            foreach (var camera in _all)
                if (camera.Actor is { IsActive: true }) return camera;

            return null;
        }
    }

    /// <inheritdoc />
    public override void Awake() => _all.Add(this);

    /// <inheritdoc />
    public override void OnDestroy() => _all.Remove(this);

    // -------------------------------------------------------------------------
    // Zoom
    // -------------------------------------------------------------------------

    private float _zoom = 1f;

    /// <summary>Current zoom level. Clamped between MinZoom and MaxZoom.</summary>
    public float Zoom
    {
        get => _zoom;
        set => _zoom = Math.Clamp(value, MinZoom, MaxZoom);
    }

    public float MinZoom { get; set; } = 0.1f;
    public float MaxZoom { get; set; } = 10f;

    // -------------------------------------------------------------------------
    // World bounds clamping
    // -------------------------------------------------------------------------

    /// <summary>
    /// Optional world-space rectangle the camera centre is clamped within.
    /// When null no clamping is applied.
    /// </summary>
    public Rectangle? Bounds { get; set; }

    // -------------------------------------------------------------------------
    // Follow behaviour
    // -------------------------------------------------------------------------

    /// <summary>Inner class managing smooth camera following of an Actor target.</summary>
    public CameraFollow Follow { get; } = new CameraFollow();

    // -------------------------------------------------------------------------
    // Shake (trauma-based)
    // -------------------------------------------------------------------------

    private float _trauma;        // 0..1, decays over shake duration
    private float _shakeDuration;
    private float _shakeTimer;

    // Random number generator for shake offsets
    private static readonly Random _rng = new Random();

    /// <summary>Current raw shake offset applied on top of the camera position.</summary>
    private Vector2 _shakeOffset;
    private float   _shakeAngleOffset;

    /// <summary>Maximum pixel displacement at full trauma.</summary>
    public float ShakeMaxOffset { get; set; } = 20f;

    /// <summary>Maximum rotational displacement (radians) at full trauma.</summary>
    public float ShakeMaxAngle { get; set; } = 0.05f;

    // -------------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------------

    public override void Update(float dt)
    {
        // Tick shake
        if (_shakeTimer > 0f)
        {
            _shakeTimer -= dt;
            _trauma = Math.Clamp(_shakeTimer / _shakeDuration, 0f, 1f);

            float shake = _trauma * _trauma; // squaring gives more punch
            _shakeOffset = new Vector2(
                ((float)_rng.NextDouble() * 2f - 1f) * ShakeMaxOffset * shake,
                ((float)_rng.NextDouble() * 2f - 1f) * ShakeMaxOffset * shake);
            _shakeAngleOffset = ((float)_rng.NextDouble() * 2f - 1f) * ShakeMaxAngle * shake;
        }
        else
        {
            _trauma = 0f;
            _shakeOffset = Vector2.Zero;
            _shakeAngleOffset = 0f;
        }
    }

    public override void LateUpdate(float dt)
    {
        Follow.Apply(this, dt);
        ClampToBounds();
    }

    // -------------------------------------------------------------------------
    // View matrix
    // -------------------------------------------------------------------------

    /// <summary>
    /// Returns the view matrix suitable for passing to SpriteBatch.Begin as the transformMatrix.
    /// Encodes camera position (with shake), rotation and zoom centred on the viewport.
    /// </summary>
    public Matrix GetViewMatrix(GraphicsDevice gd)
    {
        Vector2 camPos = Actor.Transform.Position + _shakeOffset;
        float   camRot = Actor.Transform.Rotation + _shakeAngleOffset;

        // gd.Viewport, deliberately: when a post-process pass has a render target bound this
        // is that target's size, which is the rectangle the world should be centred in.
        Vector2 viewport = new Vector2(gd.Viewport.Width, gd.Viewport.Height);

        // Translate so camera position maps to screen centre, then rotate and zoom
        return Matrix.CreateTranslation(-camPos.X, -camPos.Y, 0f)
             * Matrix.CreateRotationZ(-camRot)
             * Matrix.CreateScale(_zoom, _zoom, 1f)
             * Matrix.CreateTranslation(viewport.X * 0.5f, viewport.Y * 0.5f, 0f);
    }

    /// <summary>
    /// The world-space rectangle this camera can currently see, with a margin.
    /// </summary>
    /// <remarks>
    /// Conservative on purpose, and in three ways: the half-extent is the
    /// viewport's own diagonal rather than its width and height, so any camera
    /// rotation or shake angle is already covered; the zoom divides it; and a
    /// margin is added on top. Nothing that could appear on screen may ever fall
    /// outside this rectangle — a cull that is slightly too generous costs a few
    /// sprites, and one that is slightly too tight makes the world flicker at the
    /// edge of the window, which is far harder to recognise as a culling bug.
    /// </remarks>
    /// <param name="gd">The device whose viewport is being drawn into.</param>
    /// <param name="margin">Extra world units on every side.</param>
    public RectangleF VisibleWorldBounds(GraphicsDevice gd, float margin = 64f)
    {
        Vector2 camPos = Actor.Transform.Position + _shakeOffset;

        float halfW = gd.Viewport.Width * 0.5f;
        float halfH = gd.Viewport.Height * 0.5f;
        float reach = MathF.Sqrt(halfW * halfW + halfH * halfH) / MathF.Max(_zoom, 0.0001f) + margin;

        return new RectangleF(camPos.X - reach, camPos.Y - reach, reach * 2f, reach * 2f);
    }

    // -------------------------------------------------------------------------
    // Coordinate conversion
    // -------------------------------------------------------------------------

    /// <summary>Converts a screen-space position to world-space coordinates.</summary>
    public Vector2 ScreenToWorld(Vector2 screenPos, GraphicsDevice gd)
    {
        Matrix inv = Matrix.Invert(GetViewMatrix(gd));
        return Vector2.Transform(screenPos, inv);
    }

    /// <summary>Converts a world-space position to screen-space coordinates.</summary>
    public Vector2 WorldToScreen(Vector2 worldPos, GraphicsDevice gd)
    {
        return Vector2.Transform(worldPos, GetViewMatrix(gd));
    }

    // -------------------------------------------------------------------------
    // Shake API
    // -------------------------------------------------------------------------

    /// <summary>
    /// Trigger a screen shake.
    /// </summary>
    /// <param name="intensity">Scalar 0..1 controlling maximum displacement. Values &gt;1 are accepted but clamped in feel.</param>
    /// <param name="duration">How long (seconds) the shake lasts.</param>
    public void Shake(float intensity, float duration)
    {
        float clamped  = Math.Clamp(intensity, 0f, 1f);
        _shakeDuration = MathF.Max(duration, 0.01f);
        _shakeTimer    = _shakeDuration;
        _trauma        = clamped;
    }

    // -------------------------------------------------------------------------
    // Bounds clamping
    // -------------------------------------------------------------------------

    private void ClampToBounds()
    {
        if (Bounds == null) return;

        var pos = Actor.Transform.Position;
        pos.X = Math.Clamp(pos.X, Bounds.Value.Left, Bounds.Value.Right);
        pos.Y = Math.Clamp(pos.Y, Bounds.Value.Top,  Bounds.Value.Bottom);
        Actor.Transform.Position = pos;
    }

    // =========================================================================
    // Inner class: CameraFollow
    // =========================================================================

    /// <summary>
    /// Drives smooth camera following of a target Actor with deadzone and lerp speed.
    /// </summary>
    public sealed class CameraFollow
    {
        /// <summary>Actor the camera tracks. Null disables following.</summary>
        public Actor? Target { get; set; }

        /// <summary>
        /// Lerp speed (units per second, conceptually). Higher = snappier.
        /// The actual interpolation factor per frame is: 1 - exp(-LerpSpeed * dt).
        /// </summary>
        public float LerpSpeed { get; set; } = 5f;

        /// <summary>Constant world-space offset added to the target position before lerping.</summary>
        public Vector2 Offset { get; set; } = Vector2.Zero;

        /// <summary>
        /// Rectangular half-extents (world units) of the deadzone centred on the current camera position.
        /// The camera only starts moving when the target leaves this area.
        /// Set to Vector2.Zero to disable the deadzone.
        /// </summary>
        public Vector2 Deadzone { get; set; } = Vector2.Zero;

        /// <summary>Called by Camera2D.LateUpdate to apply the follow logic.</summary>
        internal void Apply(Camera2D camera, float dt)
        {
            if (Target == null) return;

            Vector2 targetPos = Target.Transform.Position + Offset;
            Vector2 camPos    = camera.Actor.Transform.Position;

            // Deadzone: if the target is within the deadzone, don't move
            if (Deadzone.X > 0f || Deadzone.Y > 0f)
            {
                Vector2 delta = targetPos - camPos;

                float clampedX = Math.Clamp(delta.X, -Deadzone.X, Deadzone.X);
                float clampedY = Math.Clamp(delta.Y, -Deadzone.Y, Deadzone.Y);

                // Desired position is as close as possible while keeping target inside deadzone
                targetPos = camPos + (delta - new Vector2(clampedX, clampedY));

                // If target is fully inside the deadzone, no movement needed
                if (Vector2.DistanceSquared(targetPos, camPos) < 0.01f)
                    return;
            }

            // Exponential lerp — frame-rate independent
            float t = 1f - MathF.Exp(-LerpSpeed * dt);
            camera.Actor.Transform.Position = Vector2.Lerp(camPos, targetPos, t);
        }
    }
}

File: html5/src/rendering/Camera2D.js
// -----------------------------------------------------------------------------
// Camera2D — the 2D view, mirroring Rendering/Camera2D.cs.
// -----------------------------------------------------------------------------

import { Component } from '../core/Component.js';
import { registerComponent } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { Vector2, Matrix4, SBMath } from '../math/index.js';

/** Smooth follow settings for a Camera2D. */
export class CameraFollow {
    constructor() {
        /** @type {?import('../core/Actor.js').Actor} */
        this.target = null;
        this.lerpSpeed = 5;
        this.offset = new Vector2(0, 0);
        /** Half-size of the box the target may move in before the camera reacts. */
        this.deadzone = new Vector2(0, 0);
    }
}

/** The 2D view transform. Tag its actor `MainCamera` to have the renderer find it. */
export class Camera2D extends Component {
    static schema = {
        zoom:          { type: P.Number, default: 1, min: 0.1, max: 10 },
        minZoom:       { type: P.Number, default: 0.1 },
        maxZoom:       { type: P.Number, default: 10 },
        shakeMaxOffset:{ type: P.Number, default: 20 },
        shakeMaxAngle: { type: P.Number, default: 0.05 },
    };

    /** Every live 2D camera, so the renderer can pick one without walking the scene. */
    static all = [];

    /** The camera the 2D renderer uses when none is named explicitly. */
    static get main() {
        return Camera2D.all.find((c) => c.actor?.tag === 'MainCamera')
            ?? Camera2D.all[0]
            ?? null;
    }

    constructor() {
        super();
        this._zoom = 1;
        this.minZoom = 0.1;
        this.maxZoom = 10;

        /** Optional world-space rectangle the view is confined to. */
        this.bounds = null;

        this.follow = new CameraFollow();

        this.shakeMaxOffset = 20;
        this.shakeMaxAngle = 0.05;
        this._shakeTrauma = 0;
        this._shakeDecay = 1;
        this._shakeOffset = new Vector2(0, 0);
        this._shakeAngle = 0;

        /** Viewport size in pixels. The host keeps this in step with the canvas. */
        this.viewportWidth = 1280;
        this.viewportHeight = 720;
    }

    get zoom() { return this._zoom; }
    set zoom(value) { this._zoom = SBMath.clamp(value, this.minZoom, this.maxZoom); }

    awake() { Camera2D.all.push(this); }

    onDestroy() {
        const i = Camera2D.all.indexOf(this);
        if (i >= 0) Camera2D.all.splice(i, 1);
    }

    /** Tells the camera how big the surface it is drawing to is. */
    setViewport(width, height) {
        this.viewportWidth = width;
        this.viewportHeight = height;
    }

    /**
     * The matrix the sprite batch draws through: world space into screen pixels,
     * with the camera's position at the centre of the viewport.
     */
    getViewMatrix() {
        const position = this.transform.position;
        const rotation = this.transform.rotation + this._shakeAngle;
        const halfWidth = this.viewportWidth / 2;
        const halfHeight = this.viewportHeight / 2;

        return Matrix4.multiply(
            Matrix4.multiply(
                Matrix4.multiply(
                    Matrix4.translation(
                        -position.x - this._shakeOffset.x,
                        -position.y - this._shakeOffset.y, 0),
                    Matrix4.rotationZ(-rotation)),
                Matrix4.scale(this._zoom, this._zoom, 1)),
            Matrix4.translation(halfWidth, halfHeight, 0));
    }

    /** Converts a point in canvas pixels to world space. */
    screenToWorld(screenPoint) {
        const inverse = Matrix4.invert(this.getViewMatrix());
        const p = Matrix4.transformPoint(inverse, { x: screenPoint.x, y: screenPoint.y, z: 0 });
        return new Vector2(p.x, p.y);
    }

    /** Converts a world point to canvas pixels. */
    worldToScreen(worldPoint) {
        const p = Matrix4.transformPoint(this.getViewMatrix(), { x: worldPoint.x, y: worldPoint.y, z: 0 });
        return new Vector2(p.x, p.y);
    }

    /** Adds camera shake. Intensity is 0..1; it decays over `duration` seconds. */
    shake(intensity, duration = 0.5) {
        this._shakeTrauma = Math.min(1, this._shakeTrauma + intensity);
        this._shakeDecay = duration > 0 ? 1 / duration : 1;
    }

    /**
     * Follow and shake run in lateUpdate so the camera reacts to where things
     * ended up this frame rather than where they were at the start of it.
     */
    lateUpdate(dt) {
        this._updateFollow(dt);
        this._updateShake(dt);
        this._applyBounds();
    }

    _updateFollow(dt) {
        const target = this.follow.target;
        if (!target || target.isDestroyed) return;

        const wanted = Vector2.add(target.transform.position, this.follow.offset);
        const current = this.transform.position;
        const delta = Vector2.subtract(wanted, current);

        // Inside the deadzone the camera holds still, so small idle movements do
        // not make the whole screen drift.
        if (Math.abs(delta.x) <= this.follow.deadzone.x) delta.x = 0;
        if (Math.abs(delta.y) <= this.follow.deadzone.y) delta.y = 0;

        const t = 1 - Math.exp(-this.follow.lerpSpeed * dt);
        this.transform.position = new Vector2(current.x + delta.x * t, current.y + delta.y * t);
    }

    _updateShake(dt) {
        if (this._shakeTrauma <= 0) {
            this._shakeOffset.set(0, 0);
            this._shakeAngle = 0;
            return;
        }

        this._shakeTrauma = Math.max(0, this._shakeTrauma - this._shakeDecay * dt);

        // Squaring the trauma makes a small shake feel gentle and a large one
        // violent, rather than everything reading the same.
        const magnitude = this._shakeTrauma * this._shakeTrauma;
        this._shakeOffset.set(
            (Math.random() * 2 - 1) * this.shakeMaxOffset * magnitude,
            (Math.random() * 2 - 1) * this.shakeMaxOffset * magnitude);
        this._shakeAngle = (Math.random() * 2 - 1) * this.shakeMaxAngle * magnitude;
    }

    _applyBounds() {
        if (!this.bounds) return;

        const halfWidth = this.viewportWidth / (2 * this._zoom);
        const halfHeight = this.viewportHeight / (2 * this._zoom);
        const p = this.transform.position;

        // When the view is wider than the bounds, centring beats clamping: an
        // unclampable axis would otherwise stick to one edge.
        const x = this.bounds.width <= halfWidth * 2
            ? this.bounds.x + this.bounds.width / 2
            : SBMath.clamp(p.x, this.bounds.x + halfWidth, this.bounds.right - halfWidth);
        const y = this.bounds.height <= halfHeight * 2
            ? this.bounds.y + this.bounds.height / 2
            : SBMath.clamp(p.y, this.bounds.y + halfHeight, this.bounds.bottom - halfHeight);

        this.transform.position = new Vector2(x, y);
    }
}
registerComponent(Camera2D, { category: 'Rendering', summary: 'The 2D view transform.' });
